### PR TITLE
Remove sender from map when he has no more tracked transactions

### PR DESCRIPTION
**Pull request by fab-10**
Signed-off-by: Fabio Di Fabio <fabio.difabio@consensys.net>

## PR description

In the transaction pool a sender is tracked when he has transactions, but he is never removed from the internal tracking data structure when he has no more transactions in the pool, so once a sender is added he lives forever in the pool even if he has no transactions there.
This PR remove the sender from the internal map when his last transaction is removed.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java [hunk 1]
```diff
@@ -256,8 +256,14 @@ protected void trackTransactionBySenderAndNonce(final TransactionInfo transactio
   protected void removeTransactionTrackedBySenderAndNonce(final Transaction transaction) {
     Optional.ofNullable(transactionsBySender.get(transaction.getSender()))
         .ifPresent(
-            transactionsForSender ->
-                transactionsForSender.removeTrackedTransaction(transaction.getNonce()));
+            transactionsForSender -> {
+              transactionsForSender.removeTrackedTransaction(transaction.getNonce());
+              if (transactionsForSender.transactionCount() == 0) {
+                LOG.trace(
+                    "Removing sender {} from transactionBySender since no more tracked transactions");
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java:263]: There should be an arg for this trace call, or the `{}` should be removed.

File: testutil/src/main/resources/log4j2.xml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <Configuration status="WARN">
   <Properties>
-    <Property name="root.log.level">INFO</Property>
+    <Property name="root.log.level">TRACE</Property>
```
Comment (shemnon) [inline testutil/src/main/resources/log4j2.xml:4]: This should be removed from the PR.

**Review by shemnon — CHANGES_REQUESTED** [review]

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java (hunk 1 above)
Comment (fab-10) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/sorter/AbstractPendingTransactionsSorter.java:263]: Added the argument

File: testutil/src/main/resources/log4j2.xml (hunk 2 above)
Comment (fab-10) [reply testutil/src/main/resources/log4j2.xml:4]: sure, commited by error

**Comment by garyschulte** [comment]
besu repo for testing:

https://github.com/garyschulte/besu-txparse

🍪 🍪 🍪

**Review by jflo — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -17,11 +17,12 @@
 - Create a new flag on RocksDB (_--Xplugin-rocksdb-high-spec-enabled_) for high spec hardware to boost performance
 - Transaction pool improvements to avoid filling the pool with not executable transactions, that could result in empty or semi-empty block proposals [#4425](https://github.com/hyperledger/besu/pull/4425) 
 - Limit Transaction pool consumption by sender to a configurable percentage of the pool size [#4417](https://github.com/hyperledger/besu/pull/4417)
+- Remove records that track transactions by sender when they are empty to same memory in the transaction pool [#4415](https://github.com/hyperledger/besu/pull/4415)
```
Comment (ajsutton) [inline CHANGELOG.md:20]: nit: This probably needs to move to the 22.7.4 section now.

**Review by ajsutton — APPROVED** [review]
LGTM.

File: CHANGELOG.md (hunk 3 above)
Comment (fab-10) [reply CHANGELOG.md]: moved to 22.7.4

**Pull request merged**